nixos/modules/services/networking/haproxy.nix

```diff
@@ -60,15 +60,31 @@ with lib;
       description = "HAProxy";
       after = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
+      reloadIfChanged = true;
       serviceConfig = {
         User = cfg.user;
         Group = cfg.group;
         Type = "notify";
         # when running the config test, don't be quiet so we can see what goes wrong
         ExecStartPre = "${pkgs.haproxy}/sbin/haproxy -c -f ${haproxyCfg}";
-        ExecStart = "${pkgs.haproxy}/sbin/haproxy -Ws -f ${haproxyCfg}";
-        Restart = "on-failure";
+        ExecStart = "${pkgs.haproxy}/sbin/haproxy -Ws -f ${haproxyCfg} -p /run/haproxy/haproxy.pid";
+        # support seamless reload
+        ExecReload = [
+          "${pkgs.haproxy}/sbin/haproxy -c -f ${haproxyCfg}"
+          "${pkgs.coreutils}/bin/kill -USR2 $MAINPID"
```

Comment:
How does the reload behaviour work?

Does the `"${pkgs.haproxy}/sbin/haproxy -c -f ${haproxyCfg}"` exit? Or will it become the new "main process"?

What is `$MAINPID` pointing to? Will it change after a reload?

Reply:
Yes, `"${pkgs.haproxy}/sbin/haproxy -c -f ${haproxyCfg}"` exists, it is just used to check the configuration file.

`$MAINPID` never changes, as the master process re-executes itself when receiving `$USR2`.

You're right that I missed something in the reload behaviour. The master process uses `exec(argv[0])` to re-execute itself, but that won't work as-is because the binary's path will change on updates.

I've updated the PR to make the main process execute from a symlink to the real binary (in `/run/haproxy/haproxy`).

Reply:
Hrm, I'm not sure if this brings us any further to the goal of only restarting where necessary.

I spend quite some thoughts earlier on how to handle reloads and restarts, and `reloadIfChanged` currently doesn't really work the way it should. I just wrote down in more detail an idea on how to fix this properly in https://github.com/NixOS/nixpkgs/issues/49528#issuecomment-622196884.

After something like this has landed, we could probably just make use of haproxy's "bind to a specific fd" functionality to provide seamless reloads: https://news.ycombinator.com/item?id=8004153

Reply:
> Changes of the underlying binary, change of sandboxing or other systemd options should basically always restart the service, as there's no way to apply these to an already running process.

In haproxy's case, changing the underlying binary should not restart the service, since haproxy can change it itself while preserving the connections using `expose-fd` which is already in use in this NixOS module. The last change I made (using a symlink in `/run/haproxy/haproxy` to start haproxy) allows this to work properly.

It's important to avoid haproxy restarts as much as we can, because in production this can cause significant disruption (losing all established connections), so if the changes you are proposing in #49528 are implemented, we need to expose a setting allowing `haproxy` to be reloaded even if the binary is updated.

> After something like this has landed, we could probably just make use of haproxy's "bind to a specific fd" functionality to provide seamless reloads: https://news.ycombinator.com/item?id=8004153

This is already enabled in this NixOS module (`expose-fds` is set globally), but it wasn't used before this PR since it only works with `Reload` : if the old process is killed before the new one is started, the FDs cannot be forwarded.

Indeed, `reloadIfChanged` comes with the problem that sandboxing options are never updated, and that's a problem that should be fixed.

Reply:
`haproxy` goes further than what the linked comment proposes - it doesn't yet talk about applications that support re-executing themselves with a new binary.

If we add support for something like this, generating the `/run/haproxy/haproxy` symlink in `ExecStartPre` easily won't work.
The line in `ExecStartPre` will change on a new haproxy binary (so the logic would normally decide to restart), and even if we'd add a special case for that, the symlink won't be updated if you `systemctl reload` after a new binary has been deployed.

If we do necessary changes in the module, can we make use of systemd sockets for seamless restarts, of `haproxy.service`, by making use of its `expose-fds` haproxy functionality?

Reply:
> [...] won't be updated if you systemctl reload after a new binary has been deployed.

Ah yes, but that's just because I forgot to recreate the symlink in `Reload`. I corrected this in the last push.

> If we do necessary changes in the module, can we make use of systemd sockets for seamless restarts, of haproxy.service, by making use of its expose-fds haproxy functionality?

We are already using `expose-fds` in this PR, it's used to transfer the socket between the old and new binary. Using systemd sockets should be possible, but I don't think that it is a good solution, as it would force the user to configure the bound addresses outside of haproxy's configuration.

In my opinion, the current form of this PR is the best way to allow seamless reloads. Of course, if the changes you propose in #49258 are implemented, it would need some way to indicate that the service should not be restarted even if the package is updated.

Reply:
Yes, please remove the `reloadIfChanged` for now, as it's not always working.

Please also add a small comment next to where we create the binary symlink about haproxy using `exec(argv[0])` to re-execute itself, so people understand why we create this symlink.

I assume without `reloadIfChanged`, our activation script will still just `systemctl restart haproxy.service`, right?

Reply:
Ok, I will do that, and add a comment to the exec(argv[0]) code.

> I assume without reloadIfChanged, our activation script will still just systemctl restart haproxy.service, right?

Yes.